README
Net-OAI-ORE

This module implements the OAI-ORE v1.0 specifications (see:
http://www.openarchives.org/ore/1.0/toc) to provide means 
to build and write resource maps in a number of serialization
formats. Facilities are also provided to parse some formats.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::OAI::ORE

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/Net-OAI-ORE

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-OAI-ORE

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Net-OAI-ORE

    CPAN Ratings:
        http://cpanratings.perl.org/d/Net-OAI-ORE

COPYRIGHT AND LICENCE

Copyright (C) 2007-2010 Simeon Warner

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Net-OAI-ORE

0.10    2007-10-22
        First version to match OAI-ORE alpha specification
        http://www.openarchives.org/ore/0.1/toc

$Id: Changes,v 1.1 2007-10-23 13:29:46 simeon Exp $



t/lib/arxiv_rem.pm
package arxiv_rem;

=head1 NAME

arxiv_rem - arXiv specific handling of resource map

=head1 SYNOPSIS

=cut

use strict;
use base qw(Net::OAI::ORE::ReM);
use Net::OAI::ORE::Agent;
use Net::OAI::ORE::Constant qw(:all);


=head1 METHODS

=head2 ARXIV SPECIFIC METHODS

These methods extract and return data in a format useful for ingest into arXiv.

=head3 article_title

Returns title string from resource map

=cut

sub article_title {
  my $self=shift;
  return( $self->aggregation_metadata_literal('dc:title') );
}


=head3 article_authors 

Extract author information and create arXiv format authors string.

Will die on failure

=cut

sub article_authors {
  my $self=shift;

  my @authors=$self->_sorted_authors;

  my @authorlist=();
  foreach my $author ($self->_sorted_authors) {
    my $d=$self->_author_name_mbox_affil($author);
    die "Author without name" unless ($d->{name});
    # Fix "last, first" to "first last"
    my ($last, $first)=split(',',$d->{name},2);
    $first=~s/^\s+//;
    my $entry="$first $last";
    $entry.=" (".$d->{affil}.")" if ($d->{affil});
    push(@authorlist,$entry);
  }

  return(join(', ',@authorlist)); 
}


=head3 article_abstract

Extract abstract from resource map and return string.

=cut

sub article_abstract {
  my $self=shift;
  return( $self->aggregation_metadata_literal('dcterms:abstract') );
}


=head3 article_categories

Returns string formatted category information. First is primary and any subsequent
values are secondary categories.

Will die is data cannot be extracted but does not check validity of category names
and/or combinations.

=cut

sub article_categories {
  my $self=shift;
  my $primary=$self->_primary_category;
  my $secs=$self->_secondary_categories;
  if ($secs->{$primary}) {
    delete($secs->{$primary});
  }
  return(join(' ',$primary,sort keys %$secs))
}


=head3 article_files

Returns a hashref where each entry is indexed by the file name in the 
submission package.

=cut

sub article_files {
  my $self=shift;
  return($self->_aggregated_resources_that_conform_to('http://purl.org/dc/dcmitype/Text'));
}


=head3 article_datasets

Returns a hashref where each entry is indexed by the file name in the 
submission package.

=cut

sub article_datasets {
  my $self=shift;
  return($self->_aggregated_resources_that_conform_to('http://datapub.dataconservancy.org/type/DataPub-DS'));
}


=head3 article_contact_email

Return the contact email for this article. It is taken from the first
author with an mbox specified. Any mailto: prefix will be removed.

Will die if no email can be extracted.

=cut

sub article_contact_email {
  my $self=shift;
  foreach my $author ($self->_sorted_authors) {
    my $d=$self->_author_name_mbox_affil($author);
    if ($d->{mbox}) {
      $d->{mbox}=~s/^mailto://;
      return($d->{mbox});
    }
  }
  die "Failed to extract contact email from authors listed";
}


=head2 INTERNAL METHODS

=head3 _sorted_authors

Returns an array of author resources, sorted by rank into "article order". 

Will die if no authors can be extracted.

=cut

sub _sorted_authors {
  my $self=shift;
  my @authors=$self->aggregation_metadata('dcterms:creator');
  @authors=$self->_sort_authors_by_rank(@authors);
  die "No authors!" unless (@authors);
  return(@authors);
}


=head3 _sort_authors_by_rank

Assumes that there is a rank predicate for each author and sorts the 
array or author resources by that. Will die with appropriate error
if the rank information is missing or incorrect.

=cut

sub _sort_authors_by_rank {
  my $self=shift;

  my $rankpredicate='http://datapub.dataconservancy.org/terms/rank';

  # Find rank for each author
  my %ranks=();
  foreach my $author (@_) {
    my $rankstr=$self->model->literal_matching($author,$rankpredicate);
    if (defined $rankstr) {
      my ($rank)=$rankstr=~/(\d+)/;
      if ($rankstr ne "$rank") {
        die("Bad rank number, rank='$rankstr' for ".$author->getURI);
      } elsif ($ranks{$rank}) {
        die("Multiple authors with same rank=$rank: ".$ranks{$rank}->getURI." and ".$author->getURI); 
      } else {
        $ranks{$rank}=$author;
      }
    } else {
      die("No rank for author: ".$author->getURI." (name=".$self->model->literal_matching($author,'foaf:name').")");
    }    
  }

  # Create new arrays from these ranks
  my @authors=();
  foreach my $rank (sort {$a<=>$b} keys %ranks) {
    push(@authors,$ranks{$rank});
  }

  return(@authors);
}


=head3 _author_name_mbox_affil($author)

For a single author resource $author in the resourcemap $self, find
the name, mbox and affil data and return hashref

=cut

sub _author_name_mbox_affil {
  my ($self,$author)=@_;
  
  my %data=();
  $data{name}=$self->model->literal_matching($author,'foaf:name');
  $data{mbox}=$self->model->literal_matching($author,'foaf:mbox');
  $data{affil}=$self->model->literal_matching($author,'foaf:Organization');

  return(\%data);
}


=head3 _aggregated_resources_that_conform_to($fmt)

Extract information about all aggregated resources that dcterms:conformTo
the format $fmt.

=cut

sub _aggregated_resources_that_conform_to {
  my ($self,$fmt)=@_;

  my %conforming=();
  foreach my $ar ($self->aggregated_resources) {
    foreach my $ar_fmt ($self->model->objects_matching($ar,'dcterms:conformsTo',RESOURCE)) {
      if ($ar_fmt eq $fmt) {
        my $file=$ar;
        $file=~s%^file:///?%% or die("Bad file location '$ar'");
        $conforming{$file}={};
        # look for dcterms:type and dcterms:description
        if (my $type=$self->model->literal_matching($ar,'dcterms:type')) {
          $conforming{$file}{type}=$type;
	}
        if (my $description=$self->model->literal_matching($ar,'dcterms:description')) {
          $conforming{$file}{description}=$description;
	}
      }
    }
  }
  return(\%conforming);
}


=head3 _primary_category

Extract and return the primary category. Will die if there is no primary category
or if there is more than one.

=cut

sub _primary_category {
  my $self=shift;

  my @primary=$self->model->objects_matching($self->aggregation,'http://arxiv.org/schemas/atom/primary_category',RESOURCE);
  if (scalar(@primary)==0) {
    die "Failed to extract arxiv:primary_category";
  } elsif (scalar(@primary)>1) {
    die "More than one arxiv:primary_category element (found ".scalar(@primary).")";
  }
  
  my $primary_uri=shift(@primary);
  my $primary=$self->_uri_to_category($primary_uri);
  if (not $primary) {
    die "Bad primary category URI: $primary_uri";
  }
  return($primary);
}


=head3 _secondary_categories

Extract and return the secondary categories. Does not check to see whether one of the secondaries
is a duplicate of the primary. There may be zero of more of these. Returns a hashref where each 
element is a secondary category (key) with count of occurences as value.

=cut

sub _secondary_categories {
  my $self=shift;

  my %secs=();
  foreach my $sec_uri ($self->model->objects_matching($self->aggregation,'dcterms:subject',RESOURCE)) {
    my $sec=$self->_uri_to_category($sec_uri);
    if (not $sec) {
      die "Bad secondary category URI: $sec_uri";
    }
    $secs{$sec}++;
  }
  return(\%secs);
}


=head3 _uri_to_category($uri)

Extract arXiv category name from category URI. Will return nothing if the extraction
is not possible because the URI is obviously incorrect. Does not check the category
name in detail.

=cut

sub _uri_to_category {
  my $self=shift;
  my ($uri)=@_;
  return() unless ($uri=~s%^http://arxiv.org/terms/arXiv/%%); #correct namespace
  return() unless ($uri=~/^[a-z-]+(\.[a-zA-Z-]+)$/);   #correct form
  return($uri);
}
 
1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Net::OAI::ORE',
    AUTHOR              => 'Simeon Warner <simeon@cpan.org>',
    ABSTRACT            => 'Modules to read/write/modify OAI-ORE resource maps',
    VERSION_FROM        => 'lib/Net/OAI/ORE/ReM.pm',
    PL_FILES            => {},
    PREREQ_PM => {
        'Algorithm::Diff' => 0,
	'Class::Accessor' => 0,
        'Data::UUID' => 0,
        'DateTime' => 0,
        'RDF::Core' => 0.51,
        'RDF::Core::Model' => 0,
        'RDF::Notation3' => 0.91,
        'RDF::Notation3::RDFCore' => 0,
        'Test::More' => 0,
	'XML::Atom::Syndication' => 0.942,
	'XML::Atom::Syndication::Writer' => 0,
        'XML::Writer' => 0.611,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Net-OAI-ORE-*' },
);


lib/Net/OAI/ORE/N3.pm
package Net::OAI::ORE::N3;

=head1 NAME

Net::OAI::ORE::N3

=head1 SYNPOSIS

Class to parse and serialize OAI-ORE ReMs in N3 format.

=head1 DESCRIPTION

Follows N3 specification defined by L<http://www.w3.org/DesignIssues/Notation3.html>.

=cut

use strict;
use warnings;
use Carp;

use Net::OAI::ORE::ReM;
use Net::OAI::ORE::Constant qw(:all);

use RDF::Notation3::RDFCore;
use RDF::Notation3::Triples;
use RDF::Core::Storage::Memory;
use RDF::Core::Model;
use RDF::Core::Resource;
use RDF::Core::Enumerator;
use Data::Dumper;

=head1 METHODS

=head2 new()

=cut

sub new {
  my $class=shift;
  my $self={'strict'=>'warn',
            @_};
  bless $self, (ref($class) || $class);
  return($self);
}


=head2 parse($src,$uri_rem)

Parse $src which is either a string containing the N3 serialization of 
the ReM with URI $uri_rem, or a filehandle.

=cut

sub parse {
  my $self=shift;
  my ($src,$uri_rem)=@_;  

  my $rdf = RDF::Notation3::RDFCore->new();
  $rdf->set_storage(RDF::Core::Storage::Memory->new());
  my $model;
  if (ref($src) eq 'IO::File') {
    $model = $rdf->parse_file($src); #takes IO::File or path
  } elsif (ref($src)) {
    croak("Attempt to parse N3 with a ".ref($src)." object supplied");
  } else {
    $model = $rdf->parse_string($src);
  }

  # $model is an RDF::Core::Model
  # Now we look for statements that are special to OAI-ORE
  #print "got $model ".ref($model)."\n";
  #print Dumper($model);  
  return($model);
}


=head3 serialize()

We could do this by simply converting all data to triples
and then dumping them as N3. However, we attempt to do a
nice "pretty print" specific to ORE, with a few comments for 
the parts.

=cut

sub serialize {
  my $self=shift;
  my $out='';
  my $rem=$self->{rem};
  if (ref($rem) and $rem->isa('Net::OAI::ORE::ReM')) {
    # Get the info from the ReM
    my $uri_r=$rem->uri;
    my $uri_a=$rem->aggregation;
    my @rem_and_agg=();
    my @agg_res=();
    my @lines=();
    $self->_prefixify_setup;
    foreach my $statement (@{$rem->model->as_array()}) {
      my ($subject,$predicate,$object,$is_literal)=@$statement;
      my $line=$self->_make_n3_line(@$statement);
      if ($subject eq $uri_r and $predicate eq DESCRIBES) {
        push(@rem_and_agg,$line);
      } elsif ($subject eq $uri_a and $predicate eq AGGREGATES) {
        push(@agg_res,$line);
      } else {
        push(@lines,$line);
      }
      #FIXME
    }
    # Now print it
    $out.="### OAI-ORE Resource Map ($uri_r)\n";
    $out.=$self->_prefixes_section;
    $out.="\n# Resource Map and Aggregation\n";
    $out.=join('',(sort @rem_and_agg)) if (@rem_and_agg);
    $out.="\n# Aggregated resources\n";
    $out.=join('',(sort @agg_res)) if (@agg_res);
    $out.="\n# Relations\n";
    $out.=join('',(sort @lines)) if (@lines);
    $out.="\n### End Resource Map\n";
  } else {
    carp "Can't serialize something that isn't a rem: $rem, ".ref($rem)."\n";
  }
  return($out);
}


# Create a single N3 line from subject,predicate,object labels with 
# appropriate escaping.
#
sub _make_n3_line {
  my $self=shift;
  my ($subject,$predicate,$object,$object_is_literal)=@_;
  my $line='';
  if (my $psubject=$self->_prefixify($subject)) {
    $line.=$psubject.' ';
  } else {
    $line.='<'.$subject.'> ';
  }
  if (my $ppredicate=$self->_prefixify_predicate($predicate)) {
    $line.=$ppredicate.' ';
  } else {
    $line.='<'.$predicate.'> ';
  }
  if ($object_is_literal) {
    $line.='"'._n3_escape($object).'"';
  } elsif (my $pobject=$self->_prefixify($object)) {
    $line.=$pobject;
  } else {
    $line.='<'.$object.'>';
  }
  return($line.".\n");
}


# Escape a string for use in N3 quoted string.
# See: http://www.w3.org/DesignIssues/Notation3
#
sub _n3_escape {
  my ($sstr)=@_;
  my $rstr='';
  foreach my $c (split(//,$sstr)) {
    if ($c eq '\\') {
      $c='\\\\';
    } elsif ($c eq '"') {
      $c='\\"';
    } elsif ($c eq "\n") {
      $c='\\n';
    } elsif ($c eq "\r") {
      $c='\\r';
    } elsif ($c eq "\t") {
      $c='\\t';
    }
    $rstr.=$c;
  }
  return($rstr);
}


sub _parse_warning {
  my $self=shift;
  my $msg=shift;
  print "_parse_warning: $msg\n";
}


sub _parse_error { 
  my $self=shift;
  my $msg=shift;
  if ($self->{strict} eq 'warn') {
    return($self->_parse_warning($msg));
  }
  croak "_parse_error: $msg\n";
}


# Setup ready to use _prefixify.
#
# WARNING - have to be carefull that constants do not get quoted
# in the hash. See Caveats in http://perldoc.perl.org/constant.html
#
sub _prefixify_setup {
  my $self=shift;
  $self->{prefixes_known}={
    ORE_NS() => ORE_PREFIX(),
    DC_NS() => DC_PREFIX(),
    DCT_NS() => DCT_PREFIX(),
    RDF_NS() => RDF_PREFIX() };
  $self->{prefixes_used}={};    
}


# Return prefixes section of N3 output.
#
sub _prefixes_section {
  my $self=shift;
  my $out='';
  if (scalar(keys %{$self->{prefixes_used}})) {
    $out.="\n# Namespace prefixes\n";
    foreach my $prefix (sort keys %{$self->{prefixes_used}}) {
      $out.="\@prefix $prefix: <".$self->{prefixes_used}{$prefix}.">.\n";
    }
  }
  return($out);
}


# Takes input $uri, return possible prefixed $uri having added
# prefix to list of known used prefixes.
#
sub _prefixify {
  my $self=shift;
  my ($uri)=@_;
  foreach my $prefix (keys %{$self->{prefixes_known}}) {
    if ($uri=~s%^$prefix%%) {
      $self->{prefixes_used}{$self->{prefixes_known}{$prefix}}=$prefix;
      return($self->{prefixes_known}{$prefix}.':'.$uri);
    }
  }
  # Nothing found, return nothing
  return();
}


# Special support for predicates where N3 has certain shorthands
#
sub _prefixify_predicate {
  my $self=shift;
  my ($uri)=@_;
  if ($uri eq 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type') {
    return('a');
  } elsif ($uri eq 'http://www.w3.org/2002/07/owl#sameAs') {
    return('=');
  } elsif ($uri eq 'http://www.w3.org/2000/10/swap/log#implies') {
    return('=>');
  }
  # No specials for predicate found, try normal _prefixify
  return($self->_prefixify($uri));
}


=head1 SEE ALSO

L<Net::OAI::ORE> and associated modules.

=head1 AUTHORS

Simeon Warner

=head1 LICENSE AND COPYRIGHT

Copyright 2007-2010 Simeon Warner.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=cut

1;


lib/Net/OAI/ORE/TriX.pm
package Net::OAI::ORE::TriX;

=head1 NAME

Net::OAI::ORE::TriX

=head1 SYNPOSIS

Class to parse and serialize OAI-ORE ReMs in TriX format.

=head1 DESCRIPTION

The TriX format is describe at L<http://www.w3.org/2004/03/trix/>

=cut

use strict;
use warnings;
use Carp;

use Net::OAI::ORE::ReM;
use XML::Writer;

=head1 METHODS

=head2 Creator

=head3 new(%args)

=cut

sub new {
  my $class=shift;
  my $self={@_};
  bless $self, (ref($class) || $class);
  return($self);
}

=head2 Output

=head3 serialize()

Seialize in TriX format, returns serializetion as a string.

=cut

sub serialize {
  my $self=shift;
  my $out='';
  my $rem=$self->{rem};
  if (ref($rem) and $rem->isa('Net::OAI::ORE::ReM')) {
    # Get the info from the ReM
    my $name=$rem->name();

    # create an XML writer
    my $writer=XML::Writer->new('OUTPUT'=>\$out,'DATA_MODE'=>1,'DATA_INDENT'=>1);
    $writer->startTag('trix',
                      'xmlns'=>'http://www.w3.org/2004/03/trix/trix-1/',
                      'xmlns:ore'=>'http://www.openarchives.org/ORE');
    # http://www.w3.org/2004/03/trix/trix-1/trix-1.0.xsd
    $writer->startTag('graph');

    # write name
    $writer->dataElement('uri',$name);
    
    # write metadata
    $writer->comment('required metadata');
    write_triple($writer,$name,'ore:remTimeStamp',$rem->timestamp_as_iso8601());
    write_triple($writer,$name,'ore:remAuthority',$rem->authority());

    # write aggregated resources
    $writer->comment('aggregated resources');
    foreach my $uri ($rem->aggregated_resources()) {
      write_triple($writer,$name,'ore:aggregates',$uri);
    }
   
    # write relations
    $writer->comment('rels');
    foreach my $rel (@{$rem->rels->as_array()}) {
      write_triple($writer,@$rel);
    }

    $writer->endTag('graph');
    $writer->endTag('trix');
    $writer->end();
  } else {
    carp "Can't serialize something that isn't a rem: $rem, ".ref($rem)."\n";
  }
  return($out);
}

=head2 SUBROUTINES

=head3 write_triple($writer,$subject,$predicate,$object)

Write the ($subject,$predicate,$object) to $writer.

=cut

sub write_triple {
  my ($writer,$subject,$predicate,$object)=@_;
  $writer->startTag('triple');
  # Subject and object may be id, uri, plainLiteral or typedLiteral
  $writer->dataElement('uri',$subject);
  $writer->dataElement('uri',$predicate);
  $writer->dataElement('uri',$object);
  $writer->endTag('triple');
}

=head1 SEE ALSO

L<Net::OAI::ORE::ReM>

=head1 AUTHORS

Simeon Warner

=head1 LICENSE AND COPYRIGHT

Copyright 2007-2010 Simeon Warner.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=cut

1;


lib/Net/OAI/ORE/Agent.pm
package Net::OAI::ORE::Agent;

=head1 NAME

Net::OAI::ORE::Agent - Object to represent http://purl.org/dc/terms/Agent

=head1 SYNPOSIS

  my $agent=Net::OAI::ORE::Agent->new;
  $agent->name("A Person");
  $agent->mbox("person\@example.org");
  print "Agent = ".$agent->name." <".$agent->mbox.">\n";

=head1 DESCRIPTION


=cut

use strict;
use warnings;

use Class::Accessor;

use base qw(Class::Accessor);
Net::OAI::ORE::Agent->mk_accessors(qw(uri name mbox));

=head1 METHODS

=head2 Creator 

=head3 new()

Create Net::OAI::ORE::Agent, may set uri, name and mbox via
hash arguments.

=cut

sub new {
  my $class=shift;
  my $self={@_};
  bless $self, (ref($class) || $class);
  return($self);
}

=head2 Accessors

=head3 uri

URI of the Agent. This may be a blank node id. Use instead
L<real_uri> if you want only globally meaningful URI.

=head3 real_uri

Wrapper around L<uri> which returns either a globally meaningful
URI or undef if not set or a blank node. Cannot be used to set
L<uriXX>.

=cut

sub real_uri {
  my $self=shift;
  my $uri=$self->uri(@_);
  return( $uri && $uri!~/^_/ ? $uri : undef );
}

=head3 name

Accessor for foaf:name of this Agent

=head3 mbox

Accessor for foaf:mbox of this Agent

=cut

1;


lib/Net/OAI/ORE/RDFXML.pm
package Net::OAI::ORE::RDFXML;

=head1 NAME

Net::OAI::ORE::RDFXML

=head1 SYNPOSIS

Class to parse and serialize OAI-ORE resource maps in RDF/XML format. 
See L<Net::OAI::ORE>.

=cut

use strict;
use warnings;
use Carp;

use Net::OAI::ORE::ReM;
use RDF::Core::Storage::Memory;
use RDF::Core::Model;
use RDF::Core::Model::Parser;
use RDF::Core::Model::Serializer;

=head1 METHODS

=head2 new()

Create a new RDFXML handler object.

=cut

sub new {
  my $class=shift;
  my $self={@_};
  bless $self, (ref($class) || $class);
  return($self);
}


=head2 parse($rdfxml,$uri_rem)

Parse input string or read from open filehandle $rdfxml which contains the 
RDF/XML serialization of the ReM with base URI $uri_rem. A base URI must be 
supplied otherwise the parser (L<RDF::Core::Model::Parser>) will fail.

Returns a L<RDF::Core::Model> model.

=cut

sub parse {
  my $self=shift;
  my ($rdfxml,$uri_rem)=@_;  

  # Read in file if necessary
  if (ref($rdfxml)) {
    # Assume filehandle
    my $fh=$rdfxml;
    local $/=undef; #to slurp file
    $rdfxml=<$fh>;
  }

  my $storage=RDF::Core::Storage::Memory->new();
  my $model=RDF::Core::Model->new(Storage => $storage);
  # The BaseURI must be supplied otherwise the parser will fail
  my $parser=RDF::Core::Model::Parser->new(Model=>$model,
                                           BaseURI=>$uri_rem,
                                           Source=>$rdfxml,
                                           SourceType=>'string');
  $parser->parse();
  return($model);
}


=head2 serialize()

Serialize resource map as RDF/XML. Returns serializetion
as a string.

See L<RDF::Core::Model::Serializer> and L<RDF::Core::Serializer>.

=cut

sub serialize {
  my $self=shift;
  my $out='';
  my $rem=$self->{rem};
  if (ref($rem) and $rem->isa('Net::OAI::ORE::ReM')) {
    # Get the info from the ReM
    $out.="<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
    $out.="<!-- OAI-ORE Resource Map (".$rem->uri.") -->\n";

    # Do not specify a BaseURI option as we do not want relative URIs in output
    my $serializer = RDF::Core::Model::Serializer->new(Model=>$rem->model,
                                                       Output => \$out);
    $serializer->serialize;
  } else {
    die "Can't serialize something that isn't a resource map: have a ".ref($rem)."\n";
  }
  return($out);
}


=head1 SEE ALSO

L<Net::OAI::ORE::ReM>

=head1 AUTHORS

Simeon Warner

=head1 LICENSE AND COPYRIGHT

Copyright 2007-2010 Simeon Warner.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

=cut

1;


lib/Net/OAI/ORE/Constant.pm
package Net::OAI::ORE::Constant;

use warnings;
use strict;
use Carp;

=head1 NAME

Net::OAI::ORE::Constant - Class with constants used by OAI-ORE 
Resource Map (ReM) objects

=cut

our $VERSION = '0.9';

use constant ORE_PREFIX    => 'ore';
use constant ORE_NS        => 'http://www.openarchives.org/ore/terms/';
use constant DC_PREFIX     => 'dc';
use constant DC_NS         => 'http://purl.org/dc/elements/1.1/';
use constant DCT_PREFIX    => 'dcterms';
use constant DCT_NS        => 'http://purl.org/dc/terms/';
use constant RDF_PREFIX    => 'rdf';
use constant RDF_NS        => 'http://www.w3.org/1999/02/22-rdf-syntax-ns#';
use constant FOAF_PREFIX   => 'foaf';
use constant FOAF_NS       => 'http://xmlns.com/foaf/0.1/';
use constant OWL_PREFIX    => 'owl';
use constant OWL_NS        => 'http://www.w3.org/2002/07/owl#';

use constant RESOURCE_MAP  => 'http://www.openarchives.org/ore/terms/ResourceMap';
use constant AGGREGATION   => 'http://www.openarchives.org/ore/terms/Aggregation';
use constant DESCRIBES     => 'http://www.openarchives.org/ore/terms/describes';
use constant AGGREGATES    => 'http://www.openarchives.org/ore/terms/aggregates';
use constant AGGREGATED_BY => 'http://www.openarchives.org/ore/terms/isAggregatedBy';
use constant HAS_TYPE      => 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type';
use constant CREATOR       => 'http://purl.org/dc/terms/creator';
use constant MODIFIED      => 'http://purl.org/dc/terms/modified';
use constant RIGHTS        => 'http://purl.org/dc/elements/1.1/rights';
use constant CREATED       => 'http://purl.org/dc/terms/created';
use constant FOAF_NAME     => 'http://xmlns.com/foaf/0.1/name';
use constant FOAF_MBOX     => 'http://xmlns.com/foaf/0.1/mbox';

use constant RESOURCE      => 1;
use constant LITERAL       => 2;

use constant QUIET         => 1;
use constant WARN          => 2;
use constant FATAL         => 3;
use constant RECKLESS      => 99;
use constant ERROR_LEVEL   => [ 'NO_ERROR', 'QUIET', 'WARN', 'FATAL' ]; #order must match, not values not constant

=head2 EXPORTED CONSTANTS

The following constants may be imported either individually or using
the tags show:

 ORE_PREFIX DC_PREFIX DCT_PREFIX RDF_PREFIX (tag: prefix)
 ORE_NS DC_NS DCT_NS RDF_NS (tag: ns)
 RESOURCE_MAP AGGREGATION (tag: entity)
 DESCRIBES AGGREGATES AGGREGATED_BY HAS_TYPE CREATOR CREATED RIGHTS MODIFIED (tag: rel)
 RESOURCE LITERAL (tag: internal)
 QUIET WARN FATAL RECKLESS ERROR_LEVEL (tag: err)

for example, to use MODIFIED and other relationships one might want 
to import as:

  use Net::OAI::ORE::Constant qw(:rel);
  print "The uri for modified time is ".MODIFIED."\n";

=cut

use base qw(Exporter);
our @EXPORT_OK = qw(
 ORE_PREFIX DC_PREFIX DCT_PREFIX RDF_PREFIX FOAF_PREFIX OWL_PREFIX 
 ORE_NS DC_NS DCT_NS RDF_NS FOAF_NS OWL_NS
 RESOURCE_MAP AGGREGATION
 DESCRIBES AGGREGATES AGGREGATED_BY HAS_TYPE CREATOR CREATED RIGHTS MODIFIED FOAF_NAME FOAF_MBOX
 RESOURCE LITERAL
 QUIET WARN FATAL RECKLESS ERROR_LEVEL
 namespace_and_name expand_qname
		    );
our %EXPORT_TAGS = (
 'prefix' => [qw( ORE_PREFIX DC_PREFIX DCT_PREFIX RDF_PREFIX FOAF_PREFIX OWL_PREFIX )],
 'ns' => [qw( ORE_NS DC_NS DCT_NS RDF_NS FOAF_NS OWL_NS )],
 'entity' => [qw( RESOURCE_MAP AGGREGATION )],
 'rel' => [qw( DESCRIBES AGGREGATES AGGREGATED_BY HAS_TYPE CREATOR CREATED RIGHTS MODIFIED FOAF_NAME FOAF_MBOX)],
 'internal' => [qw( RESOURCE LITERAL )],
 'err' => [qw( QUIET WARN FATAL RECKLESS ERROR_LEVEL )],
		    );

# Add all tag (easy since tags above are non-overlapping)
$EXPORT_TAGS{all}=[];
push @{$EXPORT_TAGS{all}}, @{$EXPORT_TAGS{$_}} foreach keys %EXPORT_TAGS; 
push @{$EXPORT_TAGS{all}}, 'namespace_and_name', 'expand_qname';

our %PREFIX_TO_NS = ( #parens needed to avoid perl quoting strings
  ORE_PREFIX()=>ORE_NS, DC_PREFIX()=>DC_NS, DCT_PREFIX()=>DCT_NS,
  RDF_PREFIX()=>RDF_NS, FOAF_PREFIX()=>FOAF_NS, OWL_PREFIX()=>OWL_NS );

our %NS_TO_PREFIX = ( reverse %PREFIX_TO_NS );


=head2 EXPORTED SUBROUTINES

=head3 namespace_and_name($element)

FIXME - should probably change all the constants here to use 
"Clarkian" notation for namespace and element name {ns}name.

=cut

sub namespace_and_name {
  my ($element)=@_;
  foreach my $ns (ORE_NS,DC_NS,DCT_NS,RDF_NS,FOAF_NS,OWL_NS) {
    if ($element=~/^$ns(.+)$/) {
      return($ns,$1);
    }
  }
  carp "Failed to find ns for $element";
  return('',$element);
}


=head3 expand_qname

Will attempt to expand an input resource matching the qname syntax ([a-z]+:\w+)
to a full URI if the prefix is known. Otherwise string left unchanged. It should
be safe the call this expansion on something that it either a qname or a URI
because a URI will not match the pattern AND the prefixes defined here do
not correspond to URI schemes.

See also L<http://www.w3.org/TR/REC-xml-names/#NT-QName>

=cut

sub expand_qname {
  my ($qname)=@_;
  if ($qname=~/^([a-z]+):(\w+)$/) {
    if (my $ns=$PREFIX_TO_NS{$1}) {
      $qname="$ns$2";
    }
  }
  return($qname);
}


=head1 COPYRIGHT & LICENSE

Copyright 2007-2010 Simeon Warner.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;

